std/algorithm/iteration.d

```diff
@@ -4686,6 +4688,613 @@ if (isSomeChar!C)
     }
 }
 
+// substitute
+/**
+Returns a range with all occurrences of `substs` in `r`.
+replaced with their substitution.
+
+Single value replacements (`'ö'.substitute!('ä', 'a', 'ö', 'o', 'ü', 'u)`) are
+supported as well and in $(BIGOH 1).
+
+Params:
+    r = an $(REF_ALTTEXT input range, isInputRange, std,range,primitives)
+    value = a single value which can be substituted in $(BIGOH 1)
+    substs = a set of replacements/substitutions
+
+Returns: a range with the substitutions replaced.
+
+See_Also:
+$(REF replace, std, array) for an eager replace algorithm or
+$(REF translate, std, string), and $(REF tr, std, string)
+for string algorithms with translation tables.
+*/
+template substitute(substs...)
+if (substs.length >= 2 && isExpressions!substs)
+{
+    import std.range.primitives : ElementType;
+    import std.traits : CommonType;
+
+    static assert(!(substs.length & 1), "The number of substitution parameters must be even");
+
+    /**
+      Substitute single values with compile-time substitution mappings.
+      Complexity: $(BIGOH 1) due to D's $(D switch) guaranteeing $(BIGOH 1);
+    */
+    auto substitute(Value)(Value value)
+    if (!isInputRange!Value && !is(CommonType!(Value, typeof(substs[0])) == void))
+    {
+        switch (value)
+        {
+            static foreach (i; 0 .. substs.length / 2)
+                case substs[2 * i]:
+                    return substs[2 * i + 1];
+
+            default: return value;
+        }
+    }
+
+    /// Substitute single range elements with compile-time substitution mappings
+    auto substitute(R)(R r)
+    if (isInputRange!R && !is(CommonType!(ElementType!R, typeof(substs[0])) == void))
+    {
+        return r.map!(a => substitute(a));
+    }
+
+    /// Substitute subranges with compile-time substitution mappings
+    auto substitute(R)(R r)
+    if (isInputRange!R && !is(CommonType!(ElementType!R, ElementType!(typeof(substs[0]))) == void))
+    {
+        // not implemented yet, fallback to runtime variant for now
+        return .substitute(r, substs);
+    }
+}
```

Comment:
The recent style features unified documentation and using `static if` inside the implementation to distinguish cases. The overloads have complicated constraints that make documentation confusing. For example, the user shouldn't care there are two implementations that split the case "it's an input range" vs. "it's not an input range" by means of overloads. They just care about the applicable inputs, which are easy to explain by example.